trinity/nodes/base.py

```diff
@@ -0,0 +1,76 @@
+from abc import abstractmethod
+import asyncio
+from pathlib import PurePath
+from threading import Thread
+
+from evm.chains.base import BaseChain
+from evm.db.header import BaseHeaderDB
+from p2p.service import (
+    BaseService,
+    EmptyService,
+)
+from trinity.rpc.main import (
+    RPCServer,
+)
+from trinity.rpc.ipc import (
+    IPCServer,
+)
+
+
+class Node(BaseService):
+    """
+    Create usable nodes by adding subclasses that define the following
+    unset attributes...
+    """
+
+    chain_class = None  # type: BaseChain
+    peer_chain_class = None  # type: BaseService
+
+    def __init__(
+            self,
+            headerdb: BaseHeaderDB,
+            peer_pool,
+            jsonrpc_ipc_path: PurePath = None):
+        super().__init__()
+        self._headerdb = headerdb
+        self._jsonrpc_ipc_path = jsonrpc_ipc_path
+        self._peer_pool = peer_pool
+        self._peer_chain = self.peer_chain_class(headerdb, peer_pool)
+
+    @abstractmethod
+    def get_chain(self) -> BaseChain:
+        raise NotImplementedError("Node classes must implement this method")
+
+    def make_ipc_server(self) -> IPCServer:
+        if self._jsonrpc_ipc_path:
+            rpc = RPCServer(self.get_chain())
+            return IPCServer(rpc, self._jsonrpc_ipc_path)
+        else:
+            return EmptyService()
+
+    async def _run(self):
+        ipc_server = self.make_ipc_server()
+        ipc_loop = self._make_new_loop_thread()
```

Comment:
Feels like these should *maybe* be set as properties (maybe private ones) so that they can be inspected if necessary rather than having them burried in this closure.

Reply:
I don't like the idea of weird things happening like having `ipc_server` get reassigned between run and stop, but we can add a copy of these objects on self, and use the local instance everywhere in the method to avoid the problem.